contracts/Pool.sol

```diff
@@ -391,6 +399,74 @@ contract EthPool is Pool {
+///
+/// Each of these functions can be called in any order and at any time, they have immediate effects.
+/// When all of these conditions are fulfilled, on each block the configured amount is being sent.
+/// It's extracted from the `withdraw`able balance and transferred to the receivers.
+/// The process continues automatically until the sender's balance is empty.
+///
+/// The receiver has an account, from which he can `collect` tokens sent by the senders.
+/// The available amount is updated every `cycleBlocks` blocks,
+/// so recently sent tokens may not be `collect`able immediately.
+/// `cycleBlocks` is a constant configured when the pool is deployed.
+///
+/// A single address can be used both as a sender and as a receiver.
+/// It will have 2 balances in the contract, one with tokens being sent and one with received,
+/// but with no connection between them and no shared configuration.
+/// In order to send received tokens, they must be first `collect`ed and then `topUp`ped
+/// if they are to be sent through the contract.
+///
+/// The concept of something happening periodically, e.g. every block or every `cycleBlocks` are
+/// only high-level abstractions for the user, Ethereum isn't really capable of scheduling work.
+/// The actual implementation emulates that behavior by calculating the results of the scheduled
+/// events based on how many blocks have been mined and only when a user needs their outcomes.
+///
+/// The contract assumes that all amounts in the system can be stored in signed 128-bit integers.
+/// It's guaranteed to be safe only when working with tokens with supply lower than `2 ^ 127`.
+contract Erc20Pool is Pool {
+    /// @notice The address of the ERC-20 contract which tokens the pool works with
+    IERC20 public immutable erc20;
+
+    /// @param cycleBlocks The length of cycleBlocks to be used in the contract instance.
+    /// Low values make funds more available by shortening the average duration of tokens being
+    /// frozen between being taken from senders' balances and being collectable by the receiver.
+    /// High values make collecting cheaper by making it process less cycles for a given time range.
+    /// @param _erc20 The address of an ERC-20 contract which tokens the pool will work with.
+    /// To guarantee safety the supply of the tokens must be lower than `2 ^ 127`.
+    constructor(uint64 cycleBlocks, IERC20 _erc20) public Pool(cycleBlocks) {
+        erc20 = _erc20;
+    }
+
+    /// @notice Tops up the sender balance of a sender of the message.
+    /// The contract will top up with all the tokens it has allowance for from the sender.
+    function topUp() public payable {
+        uint256 amount = erc20.allowance(msg.sender, address(this));
+        if (amount > 0) {
+            erc20.transferFrom(msg.sender, address(this), amount);
+            onTopUp(uint128(amount));
+        }
+    }
```

Comment:
In practice, this function may be problematic to use, as it'll require 2 transactions to top up a certain amount (set allowance + topup), compared to only one for the Eth version. An alternative would be to pass in the amount, and in the Eth version, reimburse any difference between `msg.value` and the amount parameter. Users can then set a high allownace *once*, and then top-up whatever they want.

Reply:
Another nit: since this function is payable, it's possible to lose money by sending ether to it. Perhaps it's a good idea to send back any ether to the sender in the erc20 version.

Reply:
Oh, I forgot to remove `payable`, fixed. I'm not trying to make both versions of `TopUp` have the same API, that's why they aren't a part of the common `Pool`. IMO they're too different to unify.

Passing an `amount` in ERC-20 `TopUp` was my initial design, but I've decided to drop it. It required you to use the desired amount twice, once for allowance and once for `TopUp` and I really couldn't find a use case where these amounts would be different.

Why would somebody allow 10 tokens, but `TopUp` with only 6? Why would he want to wait for with the other 4? If he wants to `TopUp` with 4 next month, would preparing another allowance be so hard? I'm also quite wary with dangling allowances, ANYBODY can trigger it and effectively lock the tokens, because the pool can't notice random transactions.

Reply:
Fixed